src/track-promise.cc

```diff
@@ -0,0 +1,72 @@
+#include "env.h"
+#include "env-inl.h"
+
+namespace node {
+
+using v8::Function;
+using v8::HandleScope;
+using v8::Isolate;
+using v8::Local;
+using v8::Object;
+using v8::Persistent;
+using v8::Value;
+using v8::WeakCallbackData;
+
+typedef void (*FreeCallback)(Local<Object> object, Local<Function> fn);
+
+class TrackPromise {
+ public:
+  static TrackPromise* New(Isolate* isolate, Local<Object> object);
+  inline Persistent<Object>* persistent();
+ private:
+  static void WeakCallback(const WeakCallbackData<Object, TrackPromise>&);
+  inline void WeakCallback(Isolate* isolate, Local<Object> object);
+  inline TrackPromise(Isolate* isolate, Local<Object> object);
+  ~TrackPromise();
+  Persistent<Object> persistent_;
+};
+
+
+TrackPromise* TrackPromise::New(Isolate* isolate,
+                                Local<Object> object) {
+  return new TrackPromise(isolate, object);
+}
+
+
+Persistent<Object>* TrackPromise::persistent() {
+  return &persistent_;
+}
+
+
+TrackPromise::TrackPromise(Isolate* isolate,
+                           Local<Object> object)
+    : persistent_(isolate, object) {
+  persistent_.SetWeak(this, WeakCallback);
+  persistent_.MarkIndependent();
+}
+
+
+TrackPromise::~TrackPromise() {
+  persistent_.Reset();
+}
+
+
+void TrackPromise::WeakCallback(
+    const WeakCallbackData<Object, TrackPromise>& data) {
+  data.GetParameter()->WeakCallback(data.GetIsolate(), data.GetValue());
+}
+
+
+void TrackPromise::WeakCallback(Isolate* isolate, Local<Object> object) {
+  Environment* env = Environment::GetCurrent(isolate);
+  Local<Function> fn = env->promise_unhandled_rejection();
+
+  if (!fn.IsEmpty()) {
+    HandleScope scope(isolate);
+    Local<Value> promise = object.As<Value>();
+    fn->Call(env->context(), Null(env->isolate()), 1, &promise).ToLocalChecked();
```

Comment:
Yes, you can ignore for now, but eventually you will need to decide whether you need to do anything with the value that the `fn` returns. The reason `ToLocalChecked` doesn't work is because [that function asserts](https://code.google.com/p/chromium/codesearch#chromium/src/v8/include/v8.h&l=7334&rcl=1456417974) that the `Local` be non-empty. Maybe look at using [`FromMaybe`](https://code.google.com/p/chromium/codesearch#chromium/src/v8/include/v8.h&l=377) instead.

Alternatively you can ensure that `fn` always returns non-empty.